File: Src/Mod/RSC/P32Pack01/RailVehicles/Passengers/Shoreliner/Driving Trailer/CommonScripts/CabCarEngineScript.lua
-- Engine script for the Shoreliner operated by Metro-North.
--
-- @include RollingStock/PowerSupply/Electrification.lua
-- @include RollingStock/PowerSupply/PowerSupply.lua
-- @include RollingStock/AiDirection.lua
-- @include RollingStock/BrakeLight.lua
-- @include SafetySystems/AspectDisplay/MetroNorth.lua
-- @include SafetySystems/Alerter.lua
-- @include Animation.lua
-- @include Misc.lua
-- @include RailWorks.lua
-- @include Units.lua
local adu
local alerter
local blight
local power
local aidirection
local leftwindowanim, rightwindowanim

-- Note that these are reversed compared to the P32.
local powermode = {diesel = 0, thirdrail = 1}
local messageid = {locationprobe = 10110}

Initialise = Misc.wraperrors(function()
  adu = MetroNorthAdu:new{
    getbrakesuppression = function()
      return RailWorks.GetControlValue("TrainBrakeControl", 0) >= 0.4
    end,
    getacknowledge = function()
      return RailWorks.GetControlValue("AWSReset", 0) > 0
    end,
    consistspeed_mps = 80 * Units.mph.tomps
  }

  alerter = Alerter:new{
    getacknowledge = function()
      return RailWorks.GetControlValue("AWSReset", 0) > 0
    end
  }
  alerter:start()

  blight = BrakeLight:new{}

  power = PowerSupply:new{
    modecontrol = "PowerMode",
    transition_s = 20,
    getcantransition = function() return true end,
    modes = {
      [powermode.diesel] = function(elec) return true end,
      [powermode.thirdrail] = function(elec) return true end
    },
    getautomode = function(cp)
      if cp == Electrification.autochangepoint.ai_to_thirdrail then
        return powermode.thirdrail
      elseif cp == Electrification.autochangepoint.ai_to_diesel then
        return powermode.diesel
      else
        return nil
      end
    end,
    oninit = function()
      local iselectric = power:getmode() == powermode.thirdrail
      power:setavailable(Electrification.type.thirdrail, iselectric)
    end
  }

  aidirection = AiDirection:new{}

  local doors_s = 2
  leftwindowanim = Animation:new{animation = "LeftWindow", duration_s = doors_s}
  rightwindowanim = Animation:new{
    animation = "RightWindow",
    duration_s = doors_s
  }

  RailWorks.BeginUpdate()
end)

local function setplayercontrols()
  local penalty = alerter:ispenalty() or adu:ispenalty()

  local throttle = penalty and 0 or
                     RailWorks.GetControlValue("VirtualThrottle", 0)
  RailWorks.SetControlValue("Regulator", 0, throttle)

  -- There's no virtual train brake, so just move the braking handle.
  if penalty then RailWorks.SetControlValue("TrainBrakeControl", 0, 0.6) end

  local alarm = alerter:isalarm() or adu:isalarm()
  RailWorks.SetControlValue("AWS", 0, Misc.intbool(alarm))
  RailWorks.SetControlValue("AWSWarnCount", 0, Misc.intbool(alarm))
  local alert = adu:isalertplaying()
  RailWorks.SetControlValue("SpeedIncreaseAlert", 0, Misc.intbool(alert))

  -- Match DTG's blended braking "algorithm."
  local pipepress_psi = 70 -
                          RailWorks.GetControlValue("AirBrakePipePressurePSI", 0)
  local dynbrake = math.max(pipepress_psi * 0.01428, 0)
  RailWorks.SetControlValue("DynamicBrake", 0, dynbrake)
end

local function setspeedometer()
  local speed_mph = RailWorks.GetControlValue("SpeedometerMPH", 0)
  RailWorks.SetControlValue("SpeedoHundreds", 0, Misc.getdigit(speed_mph, 2))
  RailWorks.SetControlValue("SpeedoTens", 0, Misc.getdigit(speed_mph, 1))
  RailWorks.SetControlValue("SpeedoUnits", 0, Misc.getdigit(speed_mph, 0))

  local aspect = adu:getaspect()
  local n, l, m, r
  if aspect == MetroNorthAdu.aspect.restrict then
    n, l, m, r = 0, 0, 0, 1
  elseif aspect == MetroNorthAdu.aspect.medium then
    n, l, m, r = 0, 0, 1, 0
  elseif aspect == MetroNorthAdu.aspect.limited then
    n, l, m, r = 0, 1, 0, 0
  elseif aspect == MetroNorthAdu.aspect.normal then
    n, l, m, r = 1, 0, 0, 0
  end
  RailWorks.SetControlValue("SigN", 0, n)
  RailWorks.SetControlValue("SigL", 0, l)
  RailWorks.SetControlValue("SigM", 0, m)
  RailWorks.SetControlValue("SigR", 0, r)
end

local function setcutin()
  adu:setatcstate(RailWorks.GetControlValue("ATCCutIn", 0) == 1)
  alerter:setrunstate(RailWorks.GetControlValue("ACSESCutIn", 0) == 1)
  adu:setacsesstate(false)
end

local function setcabwindows()
  RailWorks.SetTime("LeftWindow",
                    RailWorks.GetControlValue("Window Left", 0) * 2)
  RailWorks.SetTime("RightWindow",
                    RailWorks.GetControlValue("Window Right", 0) * 2)
end

local function setplayerlights()
  local headlights = RailWorks.GetControlValue("Headlights", 0)
  local showditch = headlights > 0.5 and headlights < 1.5
  RailWorks.ActivateNode("ditch_left", showditch)
  RailWorks.ActivateNode("ditch_right", showditch)

  RailWorks.ActivateNode("brakelight", blight:isapplied())

  Call("CabLight:Activate", RailWorks.GetControlValue("CabLight", 0))
end

local function sethelperlights()
  local headlights = RailWorks.GetControlValue("Headlights", 0)
  local isend = not RailWorks.Engine_SendConsistMessage(messageid.locationprobe,
                                                        "", 0)
  local showditch = headlights > 1.5 and isend
  RailWorks.ActivateNode("ditch_left", showditch)
  RailWorks.ActivateNode("ditch_right", showditch)

  RailWorks.ActivateNode("brakelight", blight:isapplied())

  Call("CabLight:Activate", RailWorks.GetControlValue("CabLight", 0))
end

local function setailights()
  local isend = not RailWorks.Engine_SendConsistMessage(messageid.locationprobe,
                                                        "", 0)
  local showditch = isend and aidirection:getdirection() ==
                      AiDirection.direction.forward
  RailWorks.ActivateNode("ditch_left", showditch)
  RailWorks.ActivateNode("ditch_right", showditch)

  local aspeed_mps = math.abs(RailWorks.GetSpeed())
  local isslow = aspeed_mps < 20 * Units.mph.tomps
  RailWorks.ActivateNode("brakelight", isslow)

  Call("CabLight:Activate", 0)
end

local function updateplayer(dt)
  adu:update(dt)
  alerter:update(dt)
  blight:playerupdate(dt)
  power:update(dt)
  leftwindowanim:update(dt)
  rightwindowanim:update(dt)

  setplayercontrols()
  setspeedometer()
  setcutin()
  setcabwindows()
  setplayerlights()
end

local function updatehelper(dt)
  power:update(dt)
  leftwindowanim:update(dt)
  rightwindowanim:update(dt)

  setcabwindows()
  sethelperlights()
end

local function updateai(dt)
  power:update(dt)
  aidirection:aiupdate(dt)
  leftwindowanim:update(dt)
  rightwindowanim:update(dt)

  setcabwindows()
  setailights()
end

Update = Misc.wraperrors(function(dt)
  if RailWorks.GetIsEngineWithKey() then
    updateplayer(dt)
  elseif RailWorks.GetIsPlayer() then
    updatehelper(dt)
  else
    updateai(dt)
  end
end)

OnControlValueChange = Misc.wraperrors(function(name, index, value)
  if name == "ExpertPowerMode" and RailWorks.GetIsEngineWithKey() and
    Misc.isinitialized() and (value == 0 or value == 1) then
    Misc.showalert("Not available in OpenNEC")
  end

  if name == "VirtualThrottle" or name == "TrainBrakeControl" then
    alerter:acknowledge()
  end

  RailWorks.SetControlValue(name, index, value)
end)

OnCustomSignalMessage = Misc.wraperrors(function(message)
  power:receivemessage(message)
  adu:receivemessage(message)
end)

OnConsistMessage = Misc.wraperrors(function(message, argument, direction)
  blight:receivemessage(message, argument, direction)

  RailWorks.Engine_SendConsistMessage(message, argument, direction)
end)

File: src/mod/Assets/RSC/P32Pack01/RailVehicles/Passengers/Shoreliner/Driving Trailer/CommonScripts/CabCarEngineScript.ts
/**
 * Metro-North Bombardier Shoreliner Cab Car
 */

import * as ale from "lib/alerter";
import * as c from "lib/constants";
import * as frp from "lib/frp";
import { FrpEngine } from "lib/frp-engine";
import { mapBehavior, rejectRepeats } from "lib/frp-extra";
import { SensedDirection } from "lib/frp-vehicle";
import * as m from "lib/math";
import { AduAspect } from "lib/nec/adu";
import * as cs from "lib/nec/cabsignals";
import * as adu from "lib/nec/twospeed-adu";
import * as ps from "lib/power-supply";
import * as rw from "lib/railworks";
import { dualModeOrder, dualModeSwitchS } from "lib/shared/p32";
import * as fx from "lib/special-fx";
import * as ui from "lib/ui";

const me = new FrpEngine(() => {
    // Dual-mode power supply
    const electrification = ps.createElectrificationBehaviorWithControlValues(me, {
        [ps.Electrification.Overhead]: "PowerOverhead",
        [ps.Electrification.ThirdRail]: "Power3rdRail",
    });
    const modeAuto = () => (me.rv.GetControlValue("ExpertPowerMode") as number) > 0.5;
    ui.createAutoPowerStatusPopup(me, modeAuto);
    // Annoyingly, the power modes are reversed for the Shoreliner. We'll use
    // the P32's order internally and reverse the player's input to the
    // Shoreliner. The downside is that this flips the mode if the player
    // switches ends.
    const modeAutoSwitch$ = ps.createDualModeAutoSwitchStream(me, ...dualModeOrder, modeAuto);
    modeAutoSwitch$(mode => {
        me.rv.SetControlValue("PowerMode", mode === ps.EngineMode.Diesel ? 0 : 1); // reversed
    });
    const modeSelect = () => {
        const cv = me.rv.GetControlValue("PowerMode") as number;
        return cv > 0.5 ? ps.EngineMode.ThirdRail : ps.EngineMode.Diesel; // reversed
    };
    const modePosition = ps.createDualModeEngineBehavior({
        e: me,
        modes: dualModeOrder,
        getPlayerMode: modeSelect,
        getAiMode: ps.EngineMode.Diesel,
        getPlayerCanSwitch: () => {
            const throttle = me.rv.GetControlValue("VirtualThrottle") as number;
            return throttle < 0.5;
        },
        transitionS: dualModeSwitchS,
        instantSwitch: modeAutoSwitch$,
        positionFromSaveOrConsist: () => (me.rv.GetControlValue("PowerStart") as number) - 1,
    });
    const setModePosition$ = frp.compose(
        me.createPlayerWithKeyUpdateStream(),
        mapBehavior(modePosition),
        rejectRepeats()
    );
    setModePosition$(position => {
        me.rv.SetControlValue("PowerStart", position + 1);
    });
    // Power3rdRail is not set correctly in the third-rail engine blueprint, so
    // set it ourselves based on the value of PowerMode.
    const fixElectrification$ = frp.compose(
        me.createFirstUpdateAfterControlsSettledStream(),
        frp.filter(resumeFromSave => !resumeFromSave),
        mapBehavior(modeSelect),
        frp.map(mode => (mode === ps.EngineMode.ThirdRail ? 1 : 0))
    );
    fixElectrification$(thirdRail => {
        me.rv.SetControlValue("Power3rdRail", thirdRail);
    });

    // Safety systems cut in/out
    const atcCutIn = () => (me.rv.GetControlValue("ATCCutIn") as number) > 0.5;
    ui.createAtcStatusPopup(me, atcCutIn);
    const alerterCutIn = () => (me.rv.GetControlValue("ACSESCutIn") as number) > 0.5;
    ui.createAlerterStatusPopup(me, alerterCutIn);

    // Safety systems and ADU
    const acknowledge = me.createAcknowledgeBehavior();
    const suppression = frp.liftN(
        (bp, lever) => bp || lever,
        me.createBrakePressureSuppressionBehavior(),
        () => (me.rv.GetControlValue("TrainBrakeControl") as number) >= 0.4
    );
    const [aduState$, aduEvents$] = adu.create({
        atc: cs.metroNorthAtc,
        e: me,
        acknowledge,
        suppression,
        atcCutIn,
        acsesCutIn: () => false,
        equipmentSpeedMps: 80 * c.mph.toMps,
        pulseCodeControlValue: "SignalSpeedLimit",
    });
    const aduStateHub$ = frp.compose(aduState$, frp.hub());
    aduStateHub$(({ aspect }) => {
        me.rv.SetControlValue("SigN", aspect === cs.FourAspect.Clear ? 1 : 0);
        me.rv.SetControlValue("SigL", aspect === cs.FourAspect.ApproachLimited ? 1 : 0);
        me.rv.SetControlValue("SigM", aspect === cs.FourAspect.Approach ? 1 : 0);
        me.rv.SetControlValue("SigR", aspect === cs.FourAspect.Restricting || aspect === AduAspect.Stop ? 1 : 0);
    });
    const aduState = frp.stepper(aduStateHub$, undefined);
    // Alerter
    const alerterState = frp.stepper(ale.create({ e: me, acknowledge, cutIn: alerterCutIn }), undefined);
    // Safety system sounds
    const upgradeEvents$ = frp.compose(
        aduEvents$,
        frp.filter(evt => evt === adu.AduEvent.Upgrade)
    );
    const upgradeSound$ = frp.compose(me.createPlayerWithKeyUpdateStream(), fx.triggerSound(1, upgradeEvents$));
    upgradeSound$(play => {
        me.rv.SetControlValue("SpeedIncreaseAlert", play ? 1 : 0);
    });
    const alarmsUpdate$ = frp.compose(
        me.createPlayerWithKeyUpdateStream(),
        mapBehavior(
            frp.liftN(
                (aduState, alerterState) =>
                    (aduState?.atcAlarm || aduState?.acsesAlarm || alerterState?.alarm) ?? false,
                aduState,
                alerterState
            )
        )
    );
    alarmsUpdate$(play => {
        me.rv.SetControlValue("AWS", play ? 1 : 0);
        me.rv.SetControlValue("AWSWarnCount", play ? 1 : 0);
    });

    // Throttle, dynamic brake, and air brake controls
    const isPenaltyBrake = frp.liftN(
        (aduState, alerterState) => (aduState?.penaltyBrake || alerterState?.penaltyBrake) ?? false,
        aduState,
        alerterState
    );
    const isPowerAvailable = frp.liftN(
        position => ps.dualModeEngineHasPower(position, ...dualModeOrder, electrification),
        modePosition
    );
    const throttle = frp.liftN(
        (isPenaltyBrake, isPowerAvailable, input) => (isPenaltyBrake || !isPowerAvailable ? 0 : input),
        isPenaltyBrake,
        isPowerAvailable,
        () => me.rv.GetControlValue("VirtualThrottle") as number
    );
    const throttle$ = frp.compose(me.createPlayerWithKeyUpdateStream(), mapBehavior(throttle));
    throttle$(v => {
        me.rv.SetControlValue("Regulator", v);
    });
    // There's no virtual train brake, so just move the braking handle.
    const fullService = 0.6;
    const setPenaltyBrake$ = frp.compose(
        me.createPlayerWithKeyUpdateStream(),
        frp.filter(_ => frp.snapshot(isPenaltyBrake)),
        frp.map(_ => fullService)
    );
    setPenaltyBrake$(v => {
        me.rv.SetControlValue("TrainBrakeControl", v);
    });
    // DTG's "blended braking" algorithm (from the P32)
    const dynamicBrake = frp.liftN(
        (modeSelect, brakePipePsi) => {
            return modeSelect === ps.EngineMode.Diesel ? (70 - brakePipePsi) * 0.01428 : 0;
        },
        modeSelect,
        () => me.rv.GetControlValue("AirBrakePipePressurePSI") as number
    );
    const dynamicBrake$ = frp.compose(me.createPlayerWithKeyUpdateStream(), mapBehavior(dynamicBrake));
    dynamicBrake$(v => {
        me.rv.SetControlValue("DynamicBrake", v);
    });

    // Ditch lights
    const areHeadLightsOn = () => {
        const cv = me.rv.GetControlValue("Headlights") as number;
        return cv > 0.5 && cv < 1.5;
    };
    const ditchLightsPlayer$ = frp.compose(me.createPlayerWithKeyUpdateStream(), mapBehavior(areHeadLightsOn));
    const ditchLightsAi$ = frp.compose(
        me.createAiUpdateStream(),
        frp.map(au => au.direction === SensedDirection.Forward)
    );
    const ditchLights$ = frp.compose(
        me.createPlayerWithoutKeyUpdateStream(),
        frp.map(_ => false),
        frp.merge(ditchLightsPlayer$),
        frp.merge(ditchLightsAi$),
        rejectRepeats()
    );
    ditchLights$(on => {
        me.rv.ActivateNode("ditch_left", on);
        me.rv.ActivateNode("ditch_right", on);
    });

    // Horn rings the bell.
    const bellControl$ = frp.compose(
        me.createOnCvChangeStreamFor("Horn"),
        frp.filter(v => v === 1),
        me.mapAutoBellStream()
    );
    bellControl$(v => {
        me.rv.SetControlValue("Bell", v);
    });

    // Speedometer
    const speedoMph$ = frp.compose(
        me.createPlayerWithKeyUpdateStream(),
        mapBehavior(me.createSpeedometerDigitsMphBehavior(3))
    );
    speedoMph$(([[h, t, u]]) => {
        me.rv.SetControlValue("SpeedoHundreds", h);
        me.rv.SetControlValue("SpeedoTens", t);
        me.rv.SetControlValue("SpeedoUnits", u);
    });

    // Cab dome light
    const cabLight = new rw.Light("CabLight");
    const cabLightPlayer$ = frp.compose(
        me.createPlayerWithKeyUpdateStream(),
        me.mapGetCvStream("CabLight"),
        frp.map(v => v > 0.5)
    );
    const cabLight$ = frp.compose(
        me.createPlayerWithoutKeyUpdateStream(),
        frp.merge(me.createAiUpdateStream()),
        frp.map(_ => false),
        frp.merge(cabLightPlayer$),
        rejectRepeats()
    );
    cabLight$(on => {
        cabLight.Activate(on);
    });

    // Passenger interior lights
    const passExteriorLights: rw.Light[] = [];
    for (let i = 0; i < 8; i++) {
        passExteriorLights.push(new rw.Light(`Carriage Light ${i + 1}`));
    }
    const passExteriorLight$ = frp.compose(
        me.createPlayerUpdateStream(),
        frp.map(_ => true),
        frp.merge(
            frp.compose(
                me.createAiUpdateStream(),
                frp.map(_ => false)
            )
        ),
        rejectRepeats()
    );
    passExteriorLight$(on => {
        passExteriorLights.forEach(light => light.Activate(on));
    });

    // Exterior windows
    const leftWindow$ = frp.compose(
        me.createPlayerWithKeyUpdateStream(),
        mapBehavior(
            frp.liftN(
                position => position * 2,
                () => me.rv.GetControlValue("Window Left") as number
            )
        ),
        rejectRepeats()
    );
    const rightWindow$ = frp.compose(
        me.createPlayerWithKeyUpdateStream(),
        mapBehavior(
            frp.liftN(
                position => position * 2,
                () => me.rv.GetControlValue("Window Right") as number
            )
        ),
        rejectRepeats()
    );
    leftWindow$(t => {
        me.rv.SetTime("LeftWindow", t);
    });
    rightWindow$(t => {
        me.rv.SetTime("RightWindow", t);
    });

    // Process OnControlValueChange events.
    const onCvChange$ = frp.compose(
        me.createOnCvChangeStream(),
        frp.reject(([name]) => name === "Bell")
    );
    onCvChange$(([name, value]) => {
        me.rv.SetControlValue(name, value);
    });

    // Set consist brake lights.
    const brakeLight$ = frp.compose(fx.createBrakeLightStreamForEngine(me), rejectRepeats());
    brakeLight$(on => {
        me.rv.ActivateNode("brakelight", on);
    });

    // Set platform door height.
    fx.setLowPlatformDoorsForEngine(me, true);

    // Enable updates.
    me.e.BeginUpdate();
});
me.setup();
